d-sandbox

<div style="text-align: center; line-height: 0; padding-top: 9px;">
  <img src="https://databricks.com/wp-content/uploads/2018/03/db-academy-rgb-1200px.png" alt="Databricks Learning" style="width: 600px; height: 163px">
</div>

# CSE Coding Assignment
## Instructions

- Please answer all questions
- You can use any language you wish (e.g. Python, Scala, SQL...)
- Several Markdown cells require completion. Please edit the Markdown cells to include your answer.
- Your final notebook should compile without errors when you click "Run All"

**Please do not publish questions. This is a confidential assignment.**

### Creating a Cluster

You will need to create a Databricks Cluster. More information on this process is available here: https://docs.databricks.com/user-guide/clusters/create.html

## Getting Started

**REQUIRED:** Run the following cells exactly as written to retrieve the necessary Coding Assignment Data Sets from Amazon S3.

In [0]:
%sh curl --remote-name-all 'https://files.training.databricks.com/assessments/cse-take-home/{covertype,kafka,treecover,u.data,u.item}.csv'

[1/5]: https://files.training.databricks.com/assessments/cse-take-home/covertype.csv --> covertype.csv

% Total % Received % Xferd Average Speed Time Time Time Current

Dload Upload Total Spent Left Speed

0 0 0 0 0 0 0 0 --:--:-- --:--:-- --:--:-- 0

100 133 100 133 0 0 2557 0 --:--:-- --:--:-- --:--:-- 2557

[2/5]: https://files.training.databricks.com/assessments/cse-take-home/kafka.csv --> kafka.csv

100 135k 100 135k 0 0 5217k 0 --:--:-- --:--:-- --:--:-- 5217k

[3/5]: https://files.training.databricks.com/assessments/cse-take-home/treecover.csv --> treecover.csv

100 642k 100 642k 0 0 27.2M 0 --:--:-- --:--:-- --:--:-- 27.2M

[4/5]: https://files.training.databricks.com/assessments/cse-take-home/u.data.csv --> u.data.csv

100 1932k 100 1932k 0 0 32.5M 0 --:--:-- --:--:-- --:--:-- 32.5M

[5/5]: https://files.training.databricks.com/assessments/cse-take-home/u.item.csv --> u.item.csv

100 230k 100 230k 0 0 28.1M 0 --:--:-- --:--:-- --:--:-- 28.1M

--_curl_--https://files.training.databricks.com/assessments/cse-take-home/covertype.csv

--_curl_--https://files.training.databricks.com/assessments/cse-take-home/kafka.csv

--_curl_--https://files.training.databricks.com/assessments/cse-take-home/treecover.csv

--_curl_--https://files.training.databricks.com/assessments/cse-take-home/u.data.csv

--_curl_--https://files.training.databricks.com/assessments/cse-take-home/u.item.csv

In [0]:
dbutils.fs.cp("file:/databricks/driver/covertype.csv", "dbfs:/FileStore/tmp/covertype.csv")
dbutils.fs.cp("file:/databricks/driver/kafka.csv", "dbfs:/FileStore/tmp/kafka.csv")
dbutils.fs.cp("file:/databricks/driver/treecover.csv", "dbfs:/FileStore/tmp/treecover.csv")
dbutils.fs.cp("file:/databricks/driver/u.data.csv", "dbfs:/FileStore/tmp/u.data.csv")
dbutils.fs.cp("file:/databricks/driver/u.item.csv", "dbfs:/FileStore/tmp/u.item.csv")

True

## Part 1: Reading and Parsing Data

### Question 1:  Code Challenge - Load a CSV

- Load the CSV file at `dbfs:/FileStore/tmp/nl/treecover.csv` into a DataFrame.
- Use Apache Spark to read in the data, assigned to the variable `treeCoverDF`.
- Your method to get the CSV file into Databricks isn't graded. We are only concerned with how you use Spark to parse and load the actual data. 
- Please use the `inferSchema` option.

In [0]:
# YOUR CODE HERE

#Reading a CSV File from dbfs FileStore

treeCoverDF=spark.read.csv('dbfs:/FileStore/tmp/treecover.csv',header=True,inferSchema=True)

In [0]:
treeCoverDF.show()

+---+---------+------+-----+--------------------------------+------------------------------+-------------------------------+----------------------------------+----------+---------+---------------+---------+
 Id|Elevation|Aspect|Slope|Horizontal_Distance_To_Hydrology|Vertical_Distance_To_Hydrology|Horizontal_Distance_To_Roadways|Horizontal_Distance_To_Fire_Points|Cover_Type|Soil_Type|Wilderness_Area|Hillshade|
+---+---------+------+-----+--------------------------------+------------------------------+-------------------------------+----------------------------------+----------+---------+---------------+---------+
 1| 2596| 51| 3| 258| 0| 510| 6279| 5| 29| 1| 9am|
 2| 2590| 56| 2| 212| -6| 390| 6225| 5| 29| 1| Noon|
 3| 2804| 139| 9| 268| 65| 3180| 6121| 2| 12| 1| 3pm|
 4| 2785| 155| 18| 242| 118| 3090| 6211| 2| 30| 1| 9am|
 5| 2595| 45| 2| 153| -1| 391| 6172| 5| 29| 1| Noon|
 6| 2579| 132| 6| 300| -15| 67| 6031| 2| 29| 1| 3pm|
 7| 2606| 45| 7| 270| 5| 633| 6256| 5| 29| 1| 9am|
 8| 2605| 49| 4| 234| 7| 573| 6228| 5| 29| 1| Noon|
 9| 2617| 45| 9| 240| 56| 666| 6244| 5| 29| 1| 3pm|
 10| 2612| 59| 10| 247| 11| 636| 6230| 5| 29| 1| 9am|
 11| 2612| 201| 4| 180| 51| 735| 6222| 5| 18| 1| Noon|
 12| 2886| 151| 11| 371| 26| 5253| 4051| 2| 30| 1| 3pm|
 13| 2742| 134| 22| 150| 69| 3215| 6091| 2| 30| 1| 9am|
 14| 2609| 214| 7| 150| 46| 771| 6211| 5| 18| 1| Noon|
 15| 2503| 157| 4| 67| 4| 674| 5600| 5| 18| 1| 3pm|
 16| 2495| 51| 7| 42| 2| 752| 5576| 5| 16| 1| 9am|
 17| 2610| 259| 1| 120| -1| 607| 6096| 5| 29| 1| Noon|
 18| 2517| 72| 7| 85| 6| 595| 5607| 5| 18| 1| 3pm|
 19| 2504| 0| 4| 95| 5| 691| 5572| 5| 18| 1| 9am|
 20| 2503| 38| 5| 85| 10| 741| 5555| 5| 18| 1| Noon|
+---+---------+------+-----+--------------------------------+------------------------------+-------------------------------+----------------------------------+----------+---------+---------------+---------+
only showing top 20 rows

### Question 2:  Code Challenge - Print the Schema

Use Apache Spark to display the Schema of the `treeCoverDF` Dataframe.

In [0]:
# YOUR CODE HERE

#treeCoverDF.printSchema()

treeCoverDF.printSchema()

root
-- Id: integer (nullable = true)
-- Elevation: integer (nullable = true)
-- Aspect: integer (nullable = true)
-- Slope: integer (nullable = true)
-- Horizontal_Distance_To_Hydrology: integer (nullable = true)
-- Vertical_Distance_To_Hydrology: integer (nullable = true)
-- Horizontal_Distance_To_Roadways: integer (nullable = true)
-- Horizontal_Distance_To_Fire_Points: integer (nullable = true)
-- Cover_Type: integer (nullable = true)
-- Soil_Type: integer (nullable = true)
-- Wilderness_Area: integer (nullable = true)
-- Hillshade: string (nullable = true)

### Question 3:  Code Challenge - Rows & Columns

Use Apache Spark to display the number of rows and columns in the DataFrame.

In [0]:
# YOUR CODE HERE

#Rows are calculated using treeCoverDF.count() it returns no of rows

#Columns are using treeCoverDF.columns it returns a List len(treeCoverDF.columns) returns the number of columns
from pyspark.sql.types import StructType,StructField,IntegerType
rows=treeCoverDF.count()
columns=len(treeCoverDF.columns)
rows_columns=[(rows,columns)]
schema_1=StructType(
  [
    StructField("ROWS",IntegerType(),False),
    StructField("COLUMNS",IntegerType(),False)
  ]
)
spark.createDataFrame(rows_columns,schema=schema_1).show()

+-----+-------+
 ROWS|COLUMNS|
+-----+-------+
15120| 12|
+-----+-------+

#Part 2: Analysis

### Question 4:  Code Challenge - Summary Statistics for a Feature

Use Apache Spark to answer these questions about the `treeCoverDF` DataFrame:
- What is the range - minimum and maximum - of values for the feature `elevation`?
- What are the mean and standard deviation of the feature `elevation`?

In [0]:
# YOUR CODE HERE

#Used Agg function on the DataFrame and calculated the Range=Maximum-Minimum
from pyspark.sql import functions as F
statisticts_treeCoverDF=treeCoverDF.agg(F.min("Elevation"),F.max("Elevation"),F.mean("Elevation"),F.stddev("Elevation"))
statisticts_treeCoverDF=statisticts_treeCoverDF.withColumnRenamed("min(Elevation)","Minimum").withColumnRenamed("max(Elevation)","Maximum").withColumnRenamed("avg(Elevation)","Mean").withColumnRenamed("stddev_samp(Elevation)","StandardDeviation")
statisticts_treeCoverDF=statisticts_treeCoverDF.withColumn("Range",statisticts_treeCoverDF.Maximum-statisticts_treeCoverDF.Minimum)
statisticts_treeCoverDF.show()

+-------+-------+------------------+------------------+-----+
Minimum|Maximum| Mean| StandardDeviation|Range|
+-------+-------+------------------+------------------+-----+
 1863| 3849|2749.3225529100528|417.67818734804985| 1986|
+-------+-------+------------------+------------------+-----+

### Answer #4:

- Min `elevation`: 1863
- Max `elevation`: 3849
- Mean `elevation`: 2749.322
- Standard Deviation of `elevation`: 417.678

### Question 5:  Code Challenge - Record Count

Use Apache Spark to answer the following question:
- How many entries in the dataset have an `elevation` greater than or equal to 2749.32 meters **AND** a `Cover_Type` of 1 or 2?

In [0]:
# YOUR CODE HERE
#First Evaluates treeCoverDF["Elevation"] > = 2749.32
#Second Evaluates treeCoverDF["Cover_Type"] == 1
#third Evaluates treeCoverDF["Cover_Type"] == 2
#Evaluates all combined results
#Returns no of records that matches
treeCoverDF.filter((treeCoverDF["Elevation"] >= 2749.32) & ((treeCoverDF["Cover_Type"]==1) | (treeCoverDF["Cover_Type"]==2))).count()


3883

### Question 6: Code Challenge - Compute a Percentage

Use Apache Spark to answer the following question:
- What percentage of entries with `Cover_Type` 1 or 2 have an `elevation` at or above 2749.32 meters?

In [0]:
# YOUR CODE HERE

# (Condition Statisfied)/(Total_Entries)*100

condition_satisfied=treeCoverDF.filter((treeCoverDF["Elevation"] >= 2749.32) & ((treeCoverDF["Cover_Type"] == 1) | (treeCoverDF["Cover_Type"] == 2))).count()
total_entries=treeCoverDF.count()
percent=(condition_satisfied*100)/(total_entries)
percent

25.68121693121693

### Question 7: Code Challenge - Visualize Feature Distribution

Use any [visualization tool available in the Databricks Runtime](https://docs.databricks.com/user-guide/visualizations/index.html) to generate the following visualization:

- a bar chart that helps visualize the distribution of different Wilderness Areas in our dataset

In [0]:
# YOUR CODE HERE
display(treeCoverDF.select('Wilderness_Area'))

Wilderness_Area 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 Showing the first 1000 rows.

### Question 8: Code Challenge - Visualize Average Elevation by Cover Type 

Use any [visualization tool available in the Databricks Runtime](https://docs.databricks.com/user-guide/visualizations/index.html) to generate the following visualization:

- a bar chart showing the average elevation of each cover type with string labels for cover type

**NOTE: you will need to match the integer values in the column `treeCoverDF.Cover_Type` to the string values in `dbfs:/FileStore/tmp/nl/covertype.csv` to retrieve the Cover Type Labels. It is recommended to use an Apache Spark join.**

In [0]:
# YOUR CODE HERE
coverType_str=spark.read.csv('dbfs:/FileStore/tmp/covertype.csv',header=True)

In [0]:
coverType_str.show()

+--------------+-----------------+
cover_type_key| cover_type_label|
+--------------+-----------------+
 1| Spruce/Fir|
 2| Lodgepole Pine|
 3| Ponderosa Pine|
 4|Cottonwood/Willow|
 5| Aspen|
 6| Douglas-fir|
 7| Krummholz|
+--------------+-----------------+

In [0]:
treeCover_numbers=treeCoverDF.select(treeCoverDF.Elevation,treeCoverDF.Cover_Type)
treeCover_numbers=treeCover_numbers.groupby("Cover_Type").avg("Elevation")
treeCover_numbers=treeCover_numbers.withColumnRenamed("Cover_Type","cover_type_key")
treeCover_numbers=treeCover_numbers.join(coverType_str,on="cover_type_key",how="inner")
display(treeCover_numbers.select(["cover_type_label","avg(Elevation)"]))

cover_type_label avg(Elevation) Spruce/Fir 3128.025925925926 Douglas-fir 2423.276851851852 Ponderosa Pine 2398.4231481481484 Aspen 2786.801388888889 Cottonwood/Willow 2223.42037037037 Krummholz 3362.7699074074076 Lodgepole Pine 2922.5402777777776

#Part 3: Data Ingestion, Cleansing, and Transformations

## Instructions 

This is a multi-step, data pipeline question in which you need to achieve a few objectives to build a successful job.

### Data Sets

#### `u.data.csv`

- The full u data set, 100000 ratings by 943 users on 1682 items. 
- Each user has rated at least 20 movies.  
- Users and items are numbered consecutively from 1. 
- The data is randomly ordered. 
- This is a tab separated file consisting of four columns: 
   - user id 
   - movie id 
   - rating 
   - date (unix seconds since 1/1/1970 UTC)

#### Desired schema

- `user_id INTEGER`
- `movie_id INTEGER`
- `rating INTEGER`
- `date DATE `

#### `u.item.csv`

- This is a `|` separated file consisting of six columns:
   - movie id
   - movie title
   - release date
   - video release date
   - IMDb URL
   - genre
- movie ids in this file match movie ids in `u.data`.

#### Desired schema

- `movie_id INTEGER`
- `movie_title STRING`

### Question 9:  Code Challenge - Load DataFrames

Use Apache Spark to perform the following:
1. define the correct schemas for each Data Set to be imported as described above  
   **note:** 
      - for `u.data.csv`, `date` *must* be stored using `DateType` with the format `yyyy-MM-dd`
      - you may need to ingest `timestamp` data using `IntegerType`
      - be sure to drop unneccesary columns for `u.item.csv`
1. import the two files as DataFrames names `uDataDF` and `uItemDF` using the schemas you defined and these paths:
   - `dbfs:/FileStore/tmp/u.data.csv`
   - `dbfs:/FileStore/tmp/u.item.csv`
1. order the `uDataDF` DataFrame by the `date` column

**NOTE:** Please display the DataFrames, `uDataDF` and `uItemDF` after loading.

#### `uDataDF`

In [0]:
# YOUR CODE HERE
from pyspark.sql.types import StructType,StructField,DateType,IntegerType
from pyspark.sql import functions as F

schema_data=StructType(
[
  StructField("user_id",IntegerType(),True),
  StructField("movie_id",IntegerType(),True),
  StructField("rating",IntegerType(),True),
  StructField("date",DateType(),True)
]
)

uDataDF=spark.read.csv('dbfs:/FileStore/tmp/u.data.csv',header=False,inferSchema=False,sep='\t')
uDataDF=uDataDF.withColumnRenamed("_c0","user_id").withColumnRenamed("_c1","movie_id").withColumnRenamed("_c2","rating").withColumnRenamed("_c3","date")
uDataDF=uDataDF.withColumn("date",F.from_unixtime(uDataDF.date,format="yyyy-MM-dd").cast(DateType()))
uDataDF=uDataDF.withColumn("user_id",uDataDF.user_id.cast(IntegerType())).withColumn("movie_id",uDataDF.movie_id.cast(IntegerType())).withColumn("rating",uDataDF.rating.cast(IntegerType()))
uDataDF.printSchema()

root
-- user_id: integer (nullable = true)
-- movie_id: integer (nullable = true)
-- rating: integer (nullable = true)
-- date: date (nullable = true)

In [0]:
uDataDF.show()

+-------+--------+------+----------+
user_id|movie_id|rating| date|
+-------+--------+------+----------+
 196| 242| 3|1997-12-04|
 186| 302| 3|1998-04-04|
 22| 377| 1|1997-11-07|
 244| 51| 2|1997-11-27|
 166| 346| 1|1998-02-02|
 298| 474| 4|1998-01-07|
 115| 265| 2|1997-12-03|
 253| 465| 5|1998-04-03|
 305| 451| 3|1998-02-01|
 6| 86| 3|1997-12-31|
 62| 257| 2|1997-11-12|
 286| 1014| 5|1997-11-17|
 200| 222| 5|1997-10-05|
 210| 40| 3|1998-03-27|
 224| 29| 3|1998-02-21|
 303| 785| 3|1997-11-14|
 122| 387| 5|1997-11-11|
 194| 274| 2|1997-11-14|
 291| 1042| 4|1997-09-21|
 234| 1184| 2|1998-04-08|
+-------+--------+------+----------+
only showing top 20 rows

In [0]:
uDataDF=uDataDF.orderBy("date")
uDataDF.show()

+-------+--------+------+----------+
user_id|movie_id|rating| date|
+-------+--------+------+----------+
 119| 382| 5|1997-09-20|
 119| 237| 5|1997-09-20|
 119| 718| 5|1997-09-20|
 276| 40| 3|1997-09-20|
 276| 796| 1|1997-09-20|
 276| 70| 4|1997-09-20|
 119| 1153| 5|1997-09-20|
 276| 417| 4|1997-09-20|
 119| 222| 5|1997-09-20|
 276| 188| 4|1997-09-20|
 259| 255| 4|1997-09-20|
 276| 64| 5|1997-09-20|
 276| 142| 3|1997-09-20|
 276| 181| 5|1997-09-20|
 23| 323| 2|1997-09-20|
 276| 294| 4|1997-09-20|
 23| 196| 2|1997-09-20|
 276| 324| 4|1997-09-20|
 276| 564| 3|1997-09-20|
 276| 419| 5|1997-09-20|
+-------+--------+------+----------+
only showing top 20 rows

#### `uItemDF`

In [0]:
# YOUR CODE HERE
uItemDF=spark.read.csv('dbfs:/FileStore/tmp/u.item.csv',header=False,sep='|')['_c0','_c1']

In [0]:
uItemDF=uItemDF.withColumnRenamed("_c0","movie_id").withColumnRenamed("_c1","movie_title")

In [0]:
uItemDF.printSchema()

root
-- movie_id: string (nullable = true)
-- movie_title: string (nullable = true)

In [0]:
uItemDF=uItemDF.withColumn("movie_id",uItemDF.movie_id.cast(IntegerType()))

In [0]:
uItemDF.printSchema()

root
-- movie_id: integer (nullable = true)
-- movie_title: string (nullable = true)

In [0]:
uItemDF.show()

+--------+--------------------+
movie_id| movie_title|
+--------+--------------------+
 1| Toy Story (1995)|
 2| GoldenEye (1995)|
 3| Four Rooms (1995)|
 4| Get Shorty (1995)|
 5| Copycat (1995)|
 6|Shanghai Triad (Y...|
 7|Twelve Monkeys (1...|
 8| Babe (1995)|
 9|Dead Man Walking ...|
 10| Richard III (1995)|
 11|Seven (Se7en) (1995)|
 12|Usual Suspects, T...|
 13|Mighty Aphrodite ...|
 14| Postino, Il (1994)|
 15|Mr. Holland's Opu...|
 16|French Twist (Gaz...|
 17|From Dusk Till Da...|
 18|White Balloon, Th...|
 19|Antonia's Line (1...|
 20|Angels and Insect...|
+--------+--------------------+
only showing top 20 rows

### Question 10:  Code Challenge - Perform a Join

Use Apache Spark to do the following:
- join `uDataDF` and `uItemDf` on `movie_id` as a new DataFrame called `uMovieDF`  
   **note:** make sure you do not create duplicate `movie_id` columns
   
**NOTE:** Please display the DataFrame `uMovieDF`.

In [0]:
# YOUR CODE HERE
uMovieDF=uDataDF.join(uItemDF,on="movie_id",how="inner")
display(uMovieDF)

movie_id user_id rating date movie_title 242 196 3 1997-12-04 Kolya (1996) 302 186 3 1998-04-04 L.A. Confidential (1997) 377 22 1 1997-11-07 Heavyweights (1994) 51 244 2 1997-11-27 Legends of the Fall (1994) 346 166 1 1998-02-02 Jackie Brown (1997) 474 298 4 1998-01-07 Dr. Strangelove or: How I Learned to Stop Worrying and Love the Bomb (1963) 265 115 2 1997-12-03 Hunt for Red October, The (1990) 465 253 5 1998-04-03 Jungle Book, The (1994) 451 305 3 1998-02-01 Grease (1978) 86 6 3 1997-12-31 Remains of the Day, The (1993) 257 62 2 1997-11-12 Men in Black (1997) 1014 286 5 1997-11-17 Romy and Michele's High School Reunion (1997) 222 200 5 1997-10-05 Star Trek: First Contact (1996) 40 210 3 1998-03-27 To Wong Foo, Thanks for Everything! Julie Newmar (1995) 29 224 3 1998-02-21 Batman Forever (1995) 785 303 3 1997-11-14 Only You (1994) 387 122 5 1997-11-11 Age of Innocence, The (1993) 274 194 2 1997-11-14 Sabrina (1995) 1042 291 4 1997-09-21 Just Cause (1995) 1184 234 2 1998-04-08 Endless Summer 2, The (1994) 392 119 4 1998-01-30 Man Without a Face, The (1993) 486 167 4 1998-04-16 Sabrina (1954) 144 299 4 1997-10-26 Die Hard (1988) 118 291 2 1997-09-21 Twister (1996) 1 308 4 1998-02-17 Toy Story (1995) 546 95 2 1997-11-10 Broken Arrow (1996) 95 38 5 1998-04-13 Aladdin (1992) 768 102 2 1998-01-02 Casper (1995) 277 63 4 1997-10-01 Restoration (1995) 234 160 5 1997-10-14 Jaws (1975) 246 50 3 1997-10-17 Chasing Amy (1997) 98 301 4 1997-12-14 Silence of the Lambs, The (1991) 193 225 4 1997-11-14 Right Stuff, The (1983) 88 290 4 1997-11-28 Sleepless in Seattle (1993) 194 97 3 1998-01-08 Sting, The (1973) 274 157 4 1998-02-07 Sabrina (1995) 1081 181 1 1997-11-08 Curdled (1996) 603 278 5 1998-03-30 Rear Window (1954) 796 276 1 1997-09-20 Speechless (1994) 32 7 4 1998-03-31 Crumb (1994) 16 10 4 1997-10-26 French Twist (Gazon maudit) (1995) 304 284 4 1998-01-20 Fly Away Home (1996) 979 201 2 1998-01-06 Trigger Effect, The (1996) 564 276 3 1997-09-20 Tales from the Hood (1995) 327 287 5 1997-09-27 Cop Land (1997) 201 246 5 1998-01-16 Evil Dead II (1987) 1137 242 5 1997-11-17 Beautiful Thing (1996) 241 249 5 1997-11-16 Last of the Mohicans, The (1992) 4 99 5 1998-02-03 Get Shorty (1995) 332 178 3 1997-12-22 Kiss the Girls (1997) 100 251 4 1998-01-31 Fargo (1996) 432 81 2 1997-10-11 Fantasia (1940) 322 260 4 1998-03-23 Murder at 1600 (1997) 181 25 5 1998-01-26 Return of the Jedi (1983) 196 59 5 1998-02-23 Dead Poets Society (1989) 679 72 2 1997-11-20 Conan the Barbarian (1981) 384 87 4 1997-11-18 Naked Gun 33 1/3: The Final Insult (1994) 143 290 5 1997-11-25 Sound of Music, The (1965) 423 42 5 1997-12-03 E.T. the Extra-Terrestrial (1982) 515 292 4 1997-12-02 Boot, Das (1981) 20 115 3 1997-12-03 Angels and Insects (1995) 288 20 1 1997-11-16 Scream (1996) 219 201 4 1998-01-06 Nightmare on Elm Street, A (1984) 526 13 3 1997-12-14 Ben-Hur (1959) 919 246 4 1998-01-16 City of Lost Children, The (1995) 26 138 5 1997-11-08 Brothers McMullen, The (1995) 232 167 1 1998-04-16 Young Guns (1988) 427 60 5 1997-12-28 To Kill a Mockingbird (1962) 304 57 5 1998-01-01 Fly Away Home (1996) 274 223 4 1998-04-02 Sabrina (1995) 512 189 4 1998-04-22 Wings of Desire (1987) 15 243 3 1997-11-20 Mr. Holland's Opus (1995) 1049 92 1 1998-03-18 House Arrest (1996) 416 246 3 1998-01-16 Old Yeller (1957) 165 194 4 1997-11-14 Jean de Florette (1986) 690 241 2 1998-02-12 Seven Years in Tibet (1997) 248 178 4 1997-12-22 Grosse Pointe Blank (1997) 1444 254 3 1998-02-03 That Darn Cat! (1965) 5 293 3 1998-03-03 Copycat (1995) 229 127 5 1998-01-09 Star Trek III: The Search for Spock (1984) 237 225 5 1997-11-14 Jerry Maguire (1996) 229 299 3 1997-10-30 Star Trek III: The Search for Spock (1984) 480 225 5 1997-11-14 North by Northwest (1959) 54 276 3 1997-09-20 Outbreak (1995) 144 291 5 1997-09-21 Die Hard (1988) 366 222 4 1997-10-30 Dangerous Minds (1995) 518 267 5 1997-11-08 Miller's Crossing (1990) 403 42 3 1997-12-03 Batman (1989) 111 11 4 1998-04-06 Truth About Cats & Dogs, The

### Question 11:  Code Challenge - Perform an Aggregation

Use Apache Spark to do the following:
1. create an aggregate DataFrame, `aggDF` by
  1. extracting the year from the `date` (of the review)
  1. getting the average rating of each film per year as a column named `average_rating`
  1. ordering descending by year and average rating
1. write the resulting dataframe to a table named "movie_by_year_average_rating" in the Default database  
   **note:** use `mode(overwrite)` 

#### Desired Schema
The schema of you resulting DataFrame should be:
- `year INTEGER`
- `movie_title STRING`
- `average_rating DOUBLE`

**NOTE:** Please display the DataFrame `aggDF`.

In [0]:
uMovieDF.columns

['movie_id', 'user_id', 'rating', 'date', 'movie_title']

In [0]:
aggDF=uMovieDF.select([F.year(uMovieDF.date),uMovieDF.movie_title,uMovieDF.rating])
aggDF=aggDF.withColumnRenamed("year(date)","year")

In [0]:
aggDF=aggDF.groupBy("year","movie_title").avg("rating")


In [0]:
aggDF=aggDF.withColumnRenamed("avg(rating)","average_rating")

In [0]:
aggDF=aggDF.orderBy(["year","average_rating"],ascending=False)

In [0]:
aggDF.createTempView("movie_by_year_average_rating")

In [0]:
aggDF.writeTo("movie_by_year_average_rating").createOrReplace()

In [0]:
aggDF.printSchema()

root
-- year: integer (nullable = true)
-- movie_title: string (nullable = true)
-- average_rating: double (nullable = true)

## Part 4: Fun with JSON

JSON values are typically passed by message brokers such as Kafka or Kinesis in a string encoding. When consumed by a Spark Structured Streaming application, this json must be converted into a nested object in order to be used.

Below is a list of json strings that represents how data might be passed from a message broker.

**Note:** Make sure to run the cell below to retrieve the sample data.

In [0]:
%python


sampleJson = [
 ('{"user":100, "ips" : ["191.168.192.101", "191.168.192.103", "191.168.192.96", "191.168.192.99"]}',), 
 ('{"user":101, "ips" : ["191.168.192.102", "191.168.192.105", "191.168.192.103", "191.168.192.107"]}',), 
 ('{"user":102, "ips" : ["191.168.192.105", "191.168.192.101", "191.168.192.105", "191.168.192.107"]}',), 
 ('{"user":103, "ips" : ["191.168.192.96", "191.168.192.100", "191.168.192.107", "191.168.192.101"]}',), 
 ('{"user":104, "ips" : ["191.168.192.99", "191.168.192.99", "191.168.192.102", "191.168.192.99"]}',), 
 ('{"user":105, "ips" : ["191.168.192.99", "191.168.192.99", "191.168.192.100", "191.168.192.96"]}',), 
]

### Question 12:  Code Challenge - Count the IPs

Use any coding techniques known to you to parse this list of JSON strings to answer the following question:
- how many occurrences of each IP address are in this list?

#### Desired Output
Your results should be this:


| ip | count |
|:-:|:-:|
| `191.168.192.96` | `3` |
| `191.168.192.99` | `6` |
| `191.168.192.100` | `2` |
| `191.168.192.101` | `3` |
| `191.168.192.102` | `2` |
| `191.168.192.103` | `2` |
| `191.168.192.105` | `3` |
| `191.168.192.107` | `3` |

**NOTE:** The order of your results is not important.

In [0]:
# YOUR CODE HERE
import json
ips_dic={}
def process(sampleJson):
  dict_obj=json.loads(sampleJson[0])
  for i in dict_obj['ips']:
    if i in ips_dic.keys():
      ips_dic[i]+=1
    else:
      ips_dic[i]=1
list(map(process,sampleJson))
final_list=[]
for k,v in ips_dic.items():
  final_list.append((k,v))
spark.createDataFrame(final_list,["ip","count"]).show()


+---------------+-----+
 ip|count|
+---------------+-----+
191.168.192.101| 3|
191.168.192.103| 2|
 191.168.192.96| 3|
 191.168.192.99| 6|
191.168.192.102| 2|
191.168.192.105| 3|
191.168.192.107| 3|
191.168.192.100| 2|
+---------------+-----+

## Part 5: The Databricks API

### Question 13: Conceptual Question - the Databricks API

In 4-5 sentences, please explain what the Databricks API is used for at a high-level.

### Answer:
Databricks API is used to perform the big data transformations by loading the data from various sources like aws s3,amazon redshift.
It is also used to load the data from Local File System and has the capability of creation of tables and directly load the data into tables.
Databricks API has by default SparkContext Object so that there is no need to explicitly specify the SparkContext Object

`EDIT THIS MARKDOWN CELL WITH YOUR REPLY`

### Question 14: Conceptual Question - Explain an API Call

In 4-5 sentences, please explain what this API call. Be sure to discuss some key attributes about the cluster.

```
$ curl -n -X POST -H 'Content-Type: application/json'                      \
  -d '{                                                                     \
  "cluster_name": "high-concurrency-cluster",                               \
  "spark_version": "4.2.x-scala2.11",                                       \
  "node_type_id": "i3.xlarge",                                              \
  "spark_conf":{                                                            \
        "spark.databricks.cluster.profile":"serverless",                    \
        "spark.databricks.repl.allowedLanguages":"sql,python,r"             \
     },                                                                     \
     "aws_attributes":{                                                     \
        "zone_id":"us-west-2c",                                             \
        "first_on_demand":1,                                                \
        "availability":"SPOT_WITH_FALLBACK",                                \
        "spot_bid_price_percent":100                                        \
     },                                                                     \
   "custom_tags":{                                                          \
        "ResourceClass":"Serverless"                                        \
     },                                                                     \
       "autoscale":{                                                        \
        "min_workers":1,                                                    \
        "max_workers":2                                                     \
     },                                                                     \
  "autotermination_minutes":10                                              \
}' https://dogfood.staging.cloud.databricks.com/api/2.0/clusters/create '
```

### Answer:
The above API defines the configuration 
cluster_name defines the name of the cluster on which spark is running
spark_version defines the version of spark used
spark.databricks.cluster.profile defines the serverless which means the application runs on serverless architecture
spark.databricks.repl.allowedLanguages the programming languages that can be used in databricks 
zone_id defines the zone_id where the Nodes and the cluster is running
autoscale based on the queries it will dynamically scale up and scale down based on requests
min_workers defines number of executors =1
max_workers defines maximum number of executors=2 when automatically scales
autotermination_minutes =10 will automatically terminate the workers execution

`EDIT THIS MARKDOWN CELL WITH YOUR REPLY`

## Part 6: Security

### Question 15: Conceptual Question - Security on Databricks

Using the Databricks Documentation, what would you recommend to a Databricks and AWS customer for **securely** storing and accessing their data.

### Answer:
AWS provides a huge number of services and when it comes to security AWS maintain IAM which means Identity Access Management which keeps track of people access policies and also has cloudtrail and many more services if there is any spike in the activity AWS keeps track of logs in order to keep the customer information safe.
`EDIT THIS MARKDOWN CELL WITH YOUR REPLY`

# This is the end of the official test. Bonus below!

## Part 7: Bonus: Data Science & Machine Learning

### Question 16: Conceptual Question - A Skewed Feature

One of these lines is the *mean* of this feature. The other is the *median*. Which of these lines is the **mean** - the red line or the black line?

<img width=400px src=https://www.evernote.com/l/AAEycL6CQ0hLi5V5pIo91Ko-Pfk2i0AnGyMB/image.png>

### Answer:
The above distribution is Assymetric Distribution.
And it is Negatively Skewed since the tail of left side is very large.
In Case of Negative Skewed Mean comes before Median.
So Red Line is Mean.
`EDIT THIS MARKDOWN CELL WITH YOUR REPLY`

### Question 17: Conceptual Question - Exploratory Data Analysis

The plots below show the distribution of home selling prices differentiated by a few categorical features. Based on these plots, **which of these categorical features** - Property Type, Exterior Quality, or Month Sold - would you expect to be most associated with Price? **Why**?

<img width=1600px src=https://www.evernote.com/l/AAHulkcc20hHSJV6D1udKiwSDCN0S6oV_5YB/image.png>

### Answer:

`EDIT THIS MARKDOWN CELL WITH YOUR REPLY`

### Question 18: Conceptual Question - Analyze Model Performance

Consider the following results for a decision tree model against training and testing data sets:

`decision tree regression - train r2 score: 0.9944`  
`decision tree regression - test r2 score:  0.3119`


What is your assessment of this model?

### Answer:
The Model is trained with the data that is overfitted because of that the Test Data accuracy is very less.
`EDIT THIS MARKDOWN CELL WITH YOUR REPLY`

### Question 19: Conceptual Question - Model Selection

A series of models has been built using the same training data, but each with a subset of features.

Consider the following results for a series of logistic regression models and then answer this question:

- Which model would you choose and why?
- What other things would you want to look at and why?

| model number | feature subset | logistic regression test accuracy|
|:-:|:-:|:-:|
| 1| feat_1 |	 0.631|
| 2| feat_2 |	 0.552|
| 3| feat_3 |	 0.868|
| 4| feat_4 |	 0.868|
| 5| feat_1, feat_2 |	 0.657|
| 6| feat_1, feat_3 |	 0.947|
| 7| feat_1, feat_4 |	 0.921|
| 8| feat_2, feat_3 |	 0.947|
| 9| feat_2, feat_4 |	 0.973|
| 10| feat_3, feat_4 |	 0.947|
| 11| feat_1, feat_2, feat_3 |	 0.947|
| 12| feat_1, feat_2, feat_4 |	 0.947|
| 13| feat_1, feat_3, feat_4 |	 0.947|
| 14| feat_2, feat_3, feat_4 |	 0.973|
| 15| feat_1, feat_2, feat_3, feat_4 |	 0.973|

### Answer:

`EDIT THIS MARKDOWN CELL WITH YOUR REPLY`

-sandbox
&copy; 2019 Databricks, Inc. All rights reserved.<br/>
Apache, Apache Spark, Spark and the Spark logo are trademarks of the <a href="http://www.apache.org/">Apache Software Foundation</a>.<br/>
<br/>
<a href="https://databricks.com/privacy-policy">Privacy Policy</a> | <a href="https://databricks.com/terms-of-use">Terms of Use</a> | <a href="http://help.databricks.com/">Support</a>